### ISSUE TITLE
converting vim-textobj-user to mini.ai

### ISSUE THREAD

**Issue by barrett-ruth** [echasnovski/mini.nvim#151]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)

### Module(s)

mini.ai

### Description

Currently trying to convert my setup to the mini ecosystem. I have several vim-textobj-user based plugins, such as [this one](https://github.com/whatyouhide/vim-textobj-xmlattr) or [this](https://github.com/kana/vim-textobj-entire).

These plugins straightforwardly define their text objet regexes in one file, but I just don't know how to convert the 
call to `textobj#user#plugin` to mini.ai.

Any help would be appreciated.

**Comment by echasnovski** [comment]
Hi there!

All custom textobjects are defined inside `config.custom_textobjects` table. [Here](https://github.com/echasnovski/mini.nvim/blob/1f20d07b1ee8b79537559ac02bef9a33440a38b5/doc/mini-ai.txt#L413) are some examples. So no need to call `textobj#user#plugin`, only define inside `require('mini.ai').setup()` call.

Here is [a help section for full specification capabilities](https://github.com/echasnovski/mini.nvim/blob/1f20d07b1ee8b79537559ac02bef9a33440a38b5/doc/mini-ai.txt#L215).

Here are [some examples of user setups](https://github.com/echasnovski/mini.nvim/discussions/36#discussioncomment-3317193).

I once quickly wrote a version of textobject for whole buffer in [this Reddit comment](https://www.reddit.com/r/neovim/comments/wa819w/comment/ilfpkbd/?utm_source=share&utm_medium=web2x&context=3).

Hope this helps.

I've been planning to do a wiki page with various setups of custom textobjects for quite a while now. Maybe in the near future.

**Comment by barrett-ruth** [comment]
Ok, most of that was what I expected, thanks for the clarification. By the way the documentation actually has an implementation of the "global" text object shorter than the one in the reddit comment [here](https://github.com/echasnovski/mini.nvim/blob/1f20d07b1ee8b79537559ac02bef9a33440a38b5/doc/mini-ai.txt#L429).

So, my last question is let's say I want to transcribe [this plugin](https://github.com/Julian/vim-textobj-variable-segment/blob/main/autoload/textobj/variable_segment.vim) to mini.ai. It looks like I should essentially copy all of the logic into a custom textobject and return the `from` and `two` properties. However the above example has differing logic for the "around" and "inside" verbs on the text object. How can I communicate these things to mini.ai in order to maintain their functionality?

**Comment by echasnovski** [comment]
Your two paragraphs are interestingly interconnected :)

Yes, there is a whole buffer textobject implementation in documentation, but it is less flexible than the one from Reddit post. The difference being that the former doesn't differ between `a` and `i` textobjects, while the latter does. The way it is done is by utilizing its input arguments: all callable custom textobject specification are called with arguments same as [`MiniAi.find_textobject()`](https://github.com/echasnovski/mini.nvim/blob/1f20d07b1ee8b79537559ac02bef9a33440a38b5/doc/mini-ai.txt#L499). There is a note about it in specs docs but I realized it is not reflected in examples.

So the shorter answer is: use first `ai_type` argument of callable specification to condition the returned region(s).

**Comment by barrett-ruth** [comment]
Sorry for the prod @echasnovski ? At this point I have constructed two regexes to simulate one vim-textobj plugin I use: one for "a," and one for "i." Is there an easy way to translate that into mini/ai?

**Comment by echasnovski** [comment]
> Sorry for the prod @echasnovski ? At this point I have constructed two regexes to simulate one vim-textobj plugin I use: one for "a," and one for "i." Is there an easy way to translate that into mini/ai?

Depends on regexes, unfortunately.

What are those textobjects?

**Comment by barrett-ruth** [comment]
The regex [here](https://github.com/whatyouhide/vim-textobj-xmlattr/blob/master/plugin/textobj/xmlattr.vim), whih correlate to:

a: `\s\+\(\([a-zA-Z0-9\-_:@.]\+\)\(=\(\(".\{-}"\)\|\(\w\+\)\)\)\=\)`
i: `\(\([a-zA-Z0-9\-_:@.]\+\)\(=\(\(".\{-}"\)\|\(\w\+\)\)\)\=\)`

**Comment by echasnovski** [comment]
> The regexex [here](https://github.com/whatyouhide/vim-textobj-xmlattr/blob/master/plugin/textobj/xmlattr.vim), which correlate to:
> 
> a: `\s\+\(\([a-zA-Z0-9\-_:@.]\+\)\(=\(\(".\{-}"\)\|\(\w\+\)\)\)\=\)` i: `\(\([a-zA-Z0-9\-_:@.]\+\)\(=\(\(".\{-}"\)\|\(\w\+\)\)\)\=\)`

That is a scary looking regex...

Do you actually want to implement xml attributes, i.e. match anything similar inside to these entries: `<div class="box" id=footer data-updatable data-content="everything"></div>`?

This can be implemented with [composed Lua pattern](https://github.com/echasnovski/mini.nvim/blob/main/lua/mini/ai.lua#L193) (as described in [textobject specification](https://github.com/echasnovski/mini.nvim/blob/1c11d5fdb71c1d766545e42b0dd99eda50f1e349/lua/mini/ai.lua#L213). Its complexity depends on how precise you want it to be.

I'd settle for simpler patterns with something like this:
```lua
require('mini.ai').setup({
  custom_textobjects = {
    x = { '%s+[^%s<>=]+=[^%s<>]+', '^%s+().*()$' },
  }
})
```

Little breakdown of what it does:
- Match these: positive whitespace + group of allowed characters + `=` + group of allowed characters. Allowed characters here are anything other than whitespace, "<", or ">". This will match most of examples, but won't match "data-updatable" (I'd suggest using built-in `aW` and `iW` for these cases).
- Treat whole match as `a` textobject, trim initial whitespace to have `i` textobject.

**Comment by barrett-ruth** [comment]
Thank you so much, my poor understanding of lua patterns is what is holding me back (although not an excuse!).

I'll look back at the documentation and try to make out a solution for this and [variable-segment](https://github.com/Julian/vim-textobj-variable-segment) in the future. My two last pesky vim plugins!

**Comment by echasnovski** [comment]
> Thank you so much, my poor understanding of lua patterns is what is holding me back (although not an excuse!).
> 
> I'll look back at the documentation and try to make out a solution for this and [variable-segment](https://github.com/Julian/vim-textobj-variable-segment) in the future. My two last pesky vim plugins!

Lua pattern are initially design to be not as complicated as regular regexes. If you at least somewhat familiar with regexes, it shouldn't be much of a problem. I mostly learned them from [here](https://www.lua.org/manual/5.1/manual.html#5.4.1) while experimenting using [built-in `string` module](https://www.lua.org/manual/5.1/manual.html#5.4). One important detail, though, is that LuaJit in Neovim also contains a so called ["frontier pattern"](http://lua-users.org/wiki/FrontierPattern) - very useful to ensure that match is as wide as necessary.

Here is a MVP for 'variable-segment'. It doesn't really work in edge cases, so you can tweak it further. And speaking of frontier pattern, it also has is a slight update for xml attributes which includes maximum whitespace to the left of it:

```lua
require('mini.ai').setup({
  custom_textobjects = {
    x = { '%f[%s]%s+[^%s<>=]+=[^%s<>]+', '^%s+().*()$' },
    v = { { '%f[%w]()%w+()_', '%f[%u]()()%w*[%l%d]()()%u' } },
  }
})
```

**Comment by barrett-ruth** [comment]
Thanks so much!

**Comment by astier** [comment]
Hi. Just wanted to ask if there is some kind of documentation/collection of mini.ai textobjects as a replacement for vim-textobj-user and nvim-various-textobjs? I would like to migrate to mini.ai from targets.vim and this two textobj-libraries but having first to reimplement every textobject seems kind of daunting.

**Comment by astier** [comment]
Thanks for letting me know about the textobjects in mini.extra. I didn't find any reference in the mini.ai-readme or docs. Maybe it should be mentioned somewhere for new users. indent-textobject is one of the most important for me. Some other objects are:

- [word-column](https://github.com/coderifous/textobj-word-column.vim)
- [variable-segment](https://github.com/Julian/vim-textobj-variable-segment)
- [key-value](https://github.com/vimtaku/vim-textobj-keyvalue)
- [url](https://github.com/mattn/vim-textobj-url)

Of course there can be many more as can be seen in [nvim-various-textobjs](https://github.com/chrisgrieser/nvim-various-textobjs) and [vim-textobj-user](https://github.com/kana/vim-textobj-user/wiki). Maybe except of word-column the objects aren't that difficult to implement by oneself but I think it would be nice to have the most common textobjects-implementatons either be mentioned in a wiki or imo even better be shipped with a separate module like mini.textobjects. Together mini.ai and mini.textobjects could replace both targets.vim and other textobject-plugins like [nvim-various-textobjs](https://github.com/chrisgrieser/nvim-various-textobjs) and [vim-textobj-user](https://github.com/kana/vim-textobj-user/wiki) and help new users to make the shift to exclusively using mini.ai.

What I find so appealing about mini.ai is that it unifies targets.vim and other textobject-plugins in a coherent way and makes it possible to use in/an, il/al, g] and g[ on all the textobjects defined by mini.ai which is not possible with 3rd-party-textobject-plugins. Moving from targets.vim to mini.ai is easy. From textobject-plugins not that much depending on how many textobjects the user is depending on.

PS: Thanks for the great plugins you are delivering. Although they are supposed to be 'mini' many of them seem to be equal or better than the old established 'big' plugins.

**Comment by echasnovski** [comment]
> I didn't find any reference in the mini.ai-readme or docs. Maybe it should be mentioned somewhere for new users.

Indeed, thanks for noticing. [There is a mention now](https://github.com/echasnovski/mini.nvim/blob/1d4772e9f4f3ddc125a5ec1386fd4f85c7e3bb18/doc/mini-ai.txt#L30).

> * word-column

Yeah, might be tricky to implement in full. As it seems to have an extra logic for what to consider a "column".

> * variable-segment

Should be possible, but frankly I don't really like the idea. Being more explicit during editing tends to be more robust (at least for me).

> * key-value
> * url

Indeed, both seem to be possible.

------

> ... or imo even better be shipped with a separate module like mini.textobjects.

I don't think this will happen. If some textobject is a combination of being non-trivial, popular, and having concise but not trivial implementation, it might be a candidate for `MiniExtra.gen_ai_spec` entry. This is a case by case decision, which I went through based on my editing experience to land on its current entries.

I am open to suggestions. Yet, none of those four seem to fit in those criteria. I don't see concise implementation for first three and "url" seems to be rather trivial in terms of textobjects (as in "it usually can be replicated with other textobject like `iW`, `i)`, etc.).

**Comment by astier** [comment]
>Yeah, might be tricky to implement in full. As it seems to have an extra logic for what to consider a "column".

Do you mean it wouldn't be possible to implement word-column even if one would try because of mini.ai limitations?

> I don't see concise implementation for first three

Regarding variable-segment/subword why do you think it couldn't be concise if it would simply deal with camelCase. For underscores and other separators in most cases `ci_` or `da_` should be enough except for the outer subwords. Assuming we want to delete a subword with `*` being the cursor here are the cases:

- `cam*elCaseVar` -> `caseVar`
- `Cam*elCaseVar` -> `CaseVar`
- `camelCa*seVar` -> `camelVar`
- `camelCaseV*ar` -> `camelCase`

I think just covering the camelCase scenario should be concise, popular and not to trivial.

# 

For the non-camelcase scenarios an `outer-operator` combined with `ci_` and `da_` would make more sense. Like `do_` would be:

- `und*er_score_var` -> `score_var`
- `under_sc*ore_var` -> `under_score` (because `do_` would search for the next outer _)
- `under_score_v*ar` -> `under_score`
- `dol_` and `under_score_v*ar` -> `score_var`

# 

What do you think about a last-paste-object? Mapped to `g;` by nvim-various-textobjs. My main-usecase is `>g;` to indent the last pasted code.

# 
 
I found some weird behavior with mini.extra indent-object. In the following case `vai` and `vii` both only select from line 2 to 5. Shouldn't `vai` also include 1 and 6?

```
1 line
2    *line
3        line
4        line
5    line
6 line
```

In the next case `vai` and `vii` both also select lines 2 to 5. Shouldn't `vii` select 3 and 4?

```
1 line
2    line
3       *line
4        line
5    line
6 line
```

**Comment by echasnovski** [comment]
> > Yeah, might be tricky to implement in full. As it seems to have an extra logic for what to consider a "column".
> 
> Do you mean it wouldn't be possible to implement word-column even if one would try because of mini.ai limitations?

I mean that it would most certainly require a callable textobject specification with some ad hoc logic about what should be considered a "column". Definitely out of scope for 'mini.extra'.

> Regarding variable-segment/subword why do you think it couldn't be concise if it would simply deal with camelCase.

There is already [an example for camel case words](https://github.com/echasnovski/mini.nvim/blob/d7e05b30d842e543366ff5474076a0726f783a32/doc/mini-ai.txt#L323-L332). It looks scary, but it works for all four examples you listed.

> For the non-camelcase scenarios an `outer-operator` combined with `ci_` and `da_` would make more sense. Like `do_` would be:

The 'mini.ai' is about `a` (around) and `i` (inside) textobjects. Adding new one is out of scope. That said, tweaking `_` textobject to also accept space as edge should be doable via nested patterns.

> What do you think about a last-paste-object? Mapped to `g;` by nvim-various-textobjs. My main-usecase is `>g;` to indent the last pasted code.

Judging by [its source code](https://github.com/chrisgrieser/nvim-various-textobjs/blob/3ab25fe20f8f9c2035425d0cf5c9634a063cb98e/lua/various-textobjs/charwise-textobjs.lua#L334-L335), it is straight up area between `[` and `]` marks. Making it work in most cases is fairly straightforward with  the following:

```lua
local select_last_change = function()
  local mode = vim.fn.mode()
  local is_visual = mode == 'v' or mode == 'V' or mode == '\22'

  vim.api.nvim_win_set_cursor(0, vim.api.nvim_buf_get_mark(0, '['))
  vim.cmd('normal! ' .. (is_visual and 'o' or 'v'))
  vim.api.nvim_win_set_cursor(0, vim.api.nvim_buf_get_mark(0, ']'))
end
vim.keymap.set({ 'x', 'o' }, 'g;', select_last_change)
```

> I found some weird behavior with mini.extra indent-object. In the following case `vai` and `vii` both only select from line 2 to 5. Shouldn't `vai` also include 1 and 6?

At first glance, yes it should be from 1 to 6. I'll take a look later.

**Comment by echaya** [comment]
I have also recently migrated from `vim-textobj-user`. Can you provide any example that could tackle the following please?  @echasnovski 

| The 'mini.ai' is about a (around) and i (inside) textobjects. Adding new one is out of scope. That said, tweaking _ textobject to also accept space as edge should be doable via nested patterns.

I'm struggling to achieve the following 

-  `da_` or a user-defined 'daX` pattern to
    - `und*er_score_var` -> `*score_var`
    - `under_sc*ore_var` -> `under_*var` 
    - `under_score_v*ar` -> `under_score*`

- `di_` or a user-defined `diX` pattern to
    - `und*er_score_var` -> `*_score_var`
    - `under_sc*ore_var` -> `under_*_var` 
    - `under_score_v*ar` -> `under_score_*`

Many thanks in advance!

**Comment by echaya** [comment]
FWIW, below is what I settled with
``` Lua
    v = {
      {
        -- a*aa_bbb
        { "[^%w]()()[%w]+()_()" },
        -- a*aa_bb
        { "^()()[%w]+()_()" },
        -- aaa_bbb*_ccc
        { "_()()[%w]+()_()" },
        -- bbb_cc*cc
        { "()_()[%w]+()()%f[%W]" },
        -- bbb_cc*c at the end of the line
        { "()_()[%w]+()()$" },
      },
    },
```

I can't get my head's around when setting up `()` separately :)

**Comment by echasnovski** [comment]
> I can't get my head's around when setting up `()` separately :)

Here is a [description of what it does](https://github.com/echasnovski/mini.nvim/blob/8413efde2546be4dec9ea424bc6af346a14dffbe/doc/mini-ai.txt#L261-L267). It is basically a way to define how `a` and `i` textobjects will be extracted from the match.

**Comment by echaya** [comment]
> > I can't get my head's around when setting up `()` separately :)
> 
> Here is a [description of what it does](https://github.com/echasnovski/mini.nvim/blob/8413efde2546be4dec9ea424bc6af346a14dffbe/doc/mini-ai.txt#L261-L267). It is basically a way to define how `a` and `i` textobjects will be extracted from the match.

yes I read it so I came up with ` { "_()()[%w]+()_()" }` etc...but I haven't figured out how to transform it into `  { '_[^ _]+ ',    '^().().*()().$' },` format. 

Separately, as to the next step, I'm currently struggling with combining https://github.com/echasnovski/mini.nvim/issues/151#issuecomment-2336970229 with CamelCases. I was using [this](https://github.com/Julian/vim-textobj-variable-segment) before starting using `mini.ai` and it is neat to handle both CamelCase and under_score_case in one TextObj. 

Any help / suggestion will be well appreciated!

**Comment by echasnovski** [comment]
> ... but I haven't figured out how to transform it into ` { '_[^ _]+ ', '^().().*()().$' },` format.

If it works like that, then I suggest to use it like that. No need to follow a specific format. The nested pattern format was mostly a result of my previous attempts to combine several patterns at once, until it was apparent that handling them separately looked like the best choice.

> Separately, as to the next step, I'm currently struggling with combining [#151 (comment)](https://github.com/echasnovski/mini.nvim/issues/151#issuecomment-2336970229) with CamelCases. I was using [this](https://github.com/Julian/vim-textobj-variable-segment) before starting using `mini.ai` and it is neat to handle both CamelCase and under_score_case in one TextObj.
> 
> Any help / suggestion will be well appreciated!

I'd start by adding new patterns to what you already got. If this doesn't work, textobject specification [can be callable/function](https://github.com/echasnovski/mini.nvim/blob/8413efde2546be4dec9ea424bc6af346a14dffbe/doc/mini-ai.txt#L268) (either fully or instead of a single pattern) to provide vastly more flexibility (at the cost of more verbosity). In the end, it possible that 'mini.ai' nested pattern approach is not built for this kind of complex/flexible textobject specification.

Personally, it seems to be a better idea to have separate textobjects for snake and camel cases, though.
